osu.Game/Graphics/UserInterface/HistoryTextBox.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Input.Events;
+using osu.Game.Rulesets.Difficulty.Utils;
+using osuTK.Input;
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public class HistoryTextBox : FocusedTextBox
+    {
+        private readonly LimitedCapacityQueue<string> messageHistory;
+
+        public IEnumerable<string> GetMessageHistory()
+        {
+            if (HistoryCount == 0)
+                yield break;
+
+            for (int i = HistoryCount - 1; i >= 0; i--)
+                yield return messageHistory[i];
+        }
```

Comment:
This is only used for testing, and I'd argue we don't want to white-box internal state such as this (and there is no reason to do so either). Please remove this, and black-box test the components (so exercise keyboard arrow inputs on the textbox and verify that the textbox contents match your expectations).

Reply:
Removed.